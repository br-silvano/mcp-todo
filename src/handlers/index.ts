import {
  ICommandHandler,
  EchoHandler,
  HelpToolHandler,
  ListToolsHandler,
  DescribeToolHandler,
  UseToolHandler
} from './tool';

import {
  CreateTodoHandler,
  GetAllTodosHandler,
  GetTodoByIdHandler,
  UpdateTodoHandler,
  DeleteTodoHandler
} from './todo';

import { ToolsService } from '../services/ToolsService';
import { TodoService } from '../services/TodoService';
import { ToolRegistry } from '../tools/ToolRegistry';

export const buildHandlers = (
  toolsService: ToolsService,
  todoService: TodoService,
  toolRegistry: ToolRegistry
): ICommandHandler[] => [
    new EchoHandler(),
    new HelpToolHandler(toolsService),
    new ListToolsHandler(toolsService),
    new DescribeToolHandler(toolRegistry),
    new UseToolHandler(toolRegistry, todoService),
    new CreateTodoHandler(todoService),
    new GetAllTodosHandler(todoService),
    new GetTodoByIdHandler(todoService),
    new UpdateTodoHandler(todoService),
    new DeleteTodoHandler(todoService)
  ];
