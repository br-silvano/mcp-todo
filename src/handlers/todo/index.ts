import { GetAllTodosHandler } from './GetAllTodosHandler';
import { GetTodoByIdHandler } from './GetTodoByIdHandler';
import { UpdateTodoHandler } from './UpdateTodoHandler';
import { DeleteTodoHandler } from './DeleteTodoHandler';
import { CreateTodoHandler } from './CreateTodoHandler';

export {
  GetAllTodosHandler,
  GetTodoByIdHandler,
  UpdateTodoHandler,
  DeleteTodoHandler,
  CreateTodoHandler
}
