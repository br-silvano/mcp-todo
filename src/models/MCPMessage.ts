export interface MCPMessage {
  command: string;
  payload?: any;
  toolName?: string;
  apiKey: string;
}
