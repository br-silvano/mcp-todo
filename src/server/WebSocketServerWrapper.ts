import { Socket } from 'net';
import { WebSocket, WebSocketServer } from 'ws';

import { MCPMessage } from "../models/MCPMessage";
import { MCPMessageProcessor } from "../processor/MCPMessageProcessor";
import { AuthenticationService } from '../services/AuthenticationService';
import { checkAuthentication } from "../utils/authUtil";
import logger from "../utils/logger";

// Interface estendida para controle de heartbeat
export interface ExtendedWebSocket extends WebSocket {
  _socket: Socket;
  isAlive: boolean;
}

export class WebSocketServerWrapper {
  private readonly wss: WebSocketServer;
  private readonly messageProcessor: MCPMessageProcessor;
  private readonly authService: AuthenticationService;

  private readonly connectionTimeout = 5 * 60 * 1000; // 5 minutos
  private readonly heartbeatInterval = 30 * 1000; // 30 segundos
  private heartbeatIntervalId?: NodeJS.Timeout;
  private readonly blockedIps: Set<string>;

  constructor(
    port: number,
    messageProcessor: MCPMessageProcessor,
    authService: AuthenticationService,
    blockedIps: string[] = []
  ) {
    this.wss = new WebSocketServer({ port });
    this.messageProcessor = messageProcessor;
    this.authService = authService;
    this.blockedIps = new Set(blockedIps ?? []);
  }

  initialize(): void {
    logger.info("🧠 Inicializando o servidor WebSocket...");
    this.setupServer();
    this.startHeartbeat();
  }

  shutdown(): void {
    logger.info("🔌 Encerrando WebSocketServer...");

    this.wss.clients.forEach(client => {
      if (client.readyState === WebSocket.OPEN) {
        client.send(JSON.stringify({ type: "shutdown", reason: "Servidor encerrando." }));
      }
      client.terminate();
    });

    if (this.heartbeatIntervalId) {
      clearInterval(this.heartbeatIntervalId);
      logger.debug("🫀 Intervalo de heartbeat finalizado.");
    }

    this.wss.close((err) => {
      if (err) {
        logger.error("Erro ao encerrar o servidor WebSocket:", err);
      } else {
        logger.info("✅ Servidor WebSocket encerrado com sucesso.");
      }
    });
  }

  private setupServer(): void {
    this.wss.on('connection', (ws: ExtendedWebSocket, request) => {
      const rawIp = request.socket.remoteAddress;
      const ip = rawIp?.replace('::ffff:', '');

      if (ip && this.blockedIps.has(ip)) {
        logger.warn(`🚫 Conexão rejeitada: IP bloqueado ${ip}`);
        ws.close(1008, '🚫 Acesso não autorizado por IP');
        return;
      }

      const connectionsFromIp = [...this.wss.clients].filter(
        (client) =>
          (client as ExtendedWebSocket)._socket.remoteAddress?.replace('::ffff:', '') === ip
      );

      if (connectionsFromIp.length > 5) {
        logger.warn(`⚠️ IP ${ip} excedeu o limite de conexões simultâneas (${connectionsFromIp.length})`);
        ws.send(JSON.stringify({ error: "🔒 Limite de conexões atingido." }));
        ws.close(1013, '⚠️ Limite de conexões simultâneas excedido.');
        return;
      }

      ws.isAlive = true;
      ws.on('pong', () => {
        ws.isAlive = true;
        logger.debug(`🏓 Pong recebido de ${ip}`);
      });

      logger.info(`✅ Novo cliente conectado: ${ip}`);
      this.handleConnection(ws);
    });

    this.wss.on('listening', () => {
      logger.info(`🚀 Servidor escutando na porta ${this.wss.options.port}`);
    });

    this.wss.on('error', (error) => {
      logger.error('❌ Erro no WebSocketServer:', error);
    });
  }

  private startHeartbeat(): void {
    this.heartbeatIntervalId = setInterval(() => {
      logger.debug('🫀 Executando heartbeat...');

      this.wss.clients.forEach((client) => {
        const ws = client as ExtendedWebSocket;
        const rawIp = ws._socket.remoteAddress;
        const ip = rawIp?.replace('::ffff:', '');

        if (!ws.isAlive) {
          logger.info('⛔ Cliente inativo será encerrado.');
          return ws.terminate();
        }

        ws.isAlive = false;

        logger.debug(`🔄 Enviando ping para o cliente ${ip}`);
        ws.ping(undefined, false, (err) => {
          if (err) {
            logger.warn(`Erro ao enviar ping para ${ip}:`, err);
          } else {
            logger.debug(`Ping enviado para ${ip}`);
          }
        });
      });
    }, this.heartbeatInterval);
  }

  private handleConnection(ws: ExtendedWebSocket): void {
    const connectionStart = Date.now();
    logger.info(`📡 Conexão iniciada em ${new Date(connectionStart).toISOString()}`);

    const timeoutId = setTimeout(() => {
      if (ws.readyState === WebSocket.OPEN) {
        ws.close(1000, 'Conexão encerrada por timeout.');
        logger.info('⏱️ Conexão encerrada após 5 minutos.');
      }
    }, this.connectionTimeout);

    ws.on('message', (data: string) => this.handleMessage(ws, data));

    ws.on('close', (code, reason) => {
      const decodedReason = reason.toString() || '[sem motivo]';
      logger.info(`🔌 Conexão encerrada (code: ${code}, reason: ${decodedReason})`);
      clearTimeout(timeoutId);
    });

    ws.on('error', (error) => {
      logger.error('💥 Erro na conexão com cliente:', error);
      clearTimeout(timeoutId);
    });
  }

  private handleMessage(ws: ExtendedWebSocket, data: string): void {
    logger.info(`📨 Mensagem recebida: ${data}`);

    if (ws.readyState !== WebSocket.OPEN) {
      logger.warn('⚠️ Mensagem recebida com conexão fechada.');
      return;
    }

    let message: MCPMessage;
    try {
      message = JSON.parse(data);
    } catch (error) {
      logger.error('📛 JSON malformado:', error);
      return ws.send(JSON.stringify({ error: "Formato de mensagem inválido." }));
    }

    if (!checkAuthentication(message, this.authService)) {
      logger.warn('🔐 Autenticação falhou: chave de API inválida.');
      return ws.send(JSON.stringify({ error: "Chave de API inválida ou ausente." }));
    }

    logger.debug('🔑 Autenticado com sucesso.');
    try {
      this.messageProcessor.processMessage(message, ws);
      logger.debug('✅ Mensagem processada com sucesso.');
    } catch (error) {
      logger.error('⚙️ Erro no processamento da mensagem:', error);
    }
  }
}
