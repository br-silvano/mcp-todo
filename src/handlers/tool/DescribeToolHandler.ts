import { ICommandHandler } from "./ICommandHandler";
import { MCPMessage } from "../../models/MCPMessage";
import { ToolRegistry } from "../../tools/ToolRegistry";
import { ExtendedWebSocket } from '../../server/WebSocketServerWrapper';

export class DescribeToolHandler implements ICommandHandler {
  constructor(private readonly toolRegistry: ToolRegistry) { }

  supports(message: MCPMessage): boolean {
    return message.command === "describe-tool";
  }

  handle(message: MCPMessage, ws: ExtendedWebSocket): void {
    const { tool } = message.payload || {};
    if (!tool) {
      ws.send(JSON.stringify({ error: "Parâmetro 'tool' é obrigatório." }));
      return;
    }

    const found = this.toolRegistry.findByName(tool);
    if (!found) {
      ws.send(JSON.stringify({ error: `Tool '${tool}' não encontrada.` }));
      return;
    }

    ws.send(JSON.stringify({ tool: found }));
  }
}
