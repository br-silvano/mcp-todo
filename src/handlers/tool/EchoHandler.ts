import { MCPMessage } from "../../models/MCPMessage";
import { ICommandHandler } from "./ICommandHandler";
import { ExtendedWebSocket } from '../../server/WebSocketServerWrapper';

export class EchoHandler implements ICommandHandler {
  public supports(message: MCPMessage): boolean {
    return message.command === "echo";
  }

  public handle(message: MCPMessage, ws: ExtendedWebSocket): void {
    ws.send(JSON.stringify({
      command: "echo",
      payload: message.payload
    }));
  }
}
