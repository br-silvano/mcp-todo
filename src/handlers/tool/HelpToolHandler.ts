import { MCPMessage } from "../../models/MCPMessage";
import { ICommandHandler } from "./ICommandHandler";
import { ToolsService } from "../../services/ToolsService";
import { ExtendedWebSocket } from '../../server/WebSocketServerWrapper';

export class HelpToolHandler implements ICommandHandler {
  constructor(private readonly toolsService: ToolsService) { }

  public supports(message: MCPMessage): boolean {
    return message.command === "help-tool";
  }

  public handle(message: MCPMessage, ws: ExtendedWebSocket): void {
    const toolName = message.payload?.tool;

    if (!toolName) {
      ws.send(JSON.stringify({ error: "Nome da ferramenta não informado." }));
      return;
    }

    const tool = this.toolsService.getAvailableTools().find((t: any) => t.name === toolName);

    if (!tool) {
      ws.send(JSON.stringify({ error: `Ferramenta '${toolName}' não encontrada.` }));
      return;
    }

    ws.send(JSON.stringify({
      response: `Capacidades da ferramenta '${tool.name}':`,
      description: tool.description,
      capabilities: tool.capabilities ?? ["(sem capacidades registradas)"],
      examples: tool.examples ?? {}
    }));
  }
}
