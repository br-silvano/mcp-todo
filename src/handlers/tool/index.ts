import { ICommandHandler } from './ICommandHandler';
import { EchoHandler } from './EchoHandler';
import { HelpToolHandler } from './HelpToolHandler';
import { ListToolsHandler } from './ListToolsHandler';
import { DescribeToolHandler } from './DescribeToolHandler';
import { UseToolHandler } from './UseToolHandler';

export type {
  ICommandHandler
}

export {
  EchoHandler,
  HelpToolHandler,
  ListToolsHandler,
  DescribeToolHandler,
  UseToolHandler
}
