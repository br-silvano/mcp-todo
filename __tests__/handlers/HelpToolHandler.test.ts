import { MCPMessage } from "../../src/models/MCPMessage";
import { setupHandlers } from "../utils/setupHandlers";
import { fakeRepo } from "../utils/fakeTodoRepository";
import { MockWebSocket } from "../utils/mockWebSocket";
import { ExtendedWebSocket } from "../../src/server/WebSocketServerWrapper";
import { HelpToolHandler, ICommandHandler } from '../../src/handlers/tool';

describe("HelpToolHandler", () => {
  let handlers: ICommandHandler[];
  let mockWs: MockWebSocket;

  beforeEach(() => {
    handlers = setupHandlers(fakeRepo);
    mockWs = new MockWebSocket();
  });

  test("deve retornar os detalhes (capabilities, descrições e exemplos) da ferramenta TodoManager", () => {
    const message: MCPMessage = {
      command: "help-tool",
      apiKey: "SENHA_SECRETA",
      payload: {
        tool: "TodoManager",
      }
    };

    const helpToolHandler = handlers.find((h) => h.supports(message)) as HelpToolHandler;
    helpToolHandler.handle(message, mockWs as unknown as ExtendedWebSocket);
    const response = JSON.parse(mockWs.sentMessages[0]);

    expect(response).toMatchSnapshot();
  });
});
