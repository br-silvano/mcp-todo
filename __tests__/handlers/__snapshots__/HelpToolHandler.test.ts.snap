// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelpToolHandler deve retornar os detalhes (capabilities, descrições e exemplos) da ferramenta TodoManager 1`] = `
{
  "capabilities": [
    {
      "command": "create-todo",
      "description": "Cria uma nova tarefa com título e descrição.",
      "input_schema": {
        "properties": {
          "completed": {
            "type": "boolean",
          },
          "description": {
            "type": "string",
          },
          "title": {
            "type": "string",
          },
        },
        "required": [
          "title",
        ],
        "type": "object",
      },
      "name": "create-todo",
      "output_schema": {
        "properties": {
          "completed": {
            "type": "boolean",
          },
          "description": {
            "type": "string",
          },
          "id": {
            "type": "number",
          },
          "title": {
            "type": "string",
          },
        },
        "required": [
          "id",
          "title",
          "description",
          "completed",
        ],
        "type": "object",
      },
    },
    {
      "command": "get-todo-by-id",
      "description": "Busca uma tarefa pelo ID.",
      "input_schema": {
        "properties": {
          "id": {
            "type": "integer",
          },
        },
        "required": [
          "id",
        ],
        "type": "object",
      },
      "name": "get-todo-by-id",
      "output_schema": {
        "properties": {
          "completed": {
            "type": "boolean",
          },
          "description": {
            "type": "string",
          },
          "id": {
            "type": "integer",
          },
          "title": {
            "type": "string",
          },
        },
        "required": [
          "id",
          "title",
          "description",
          "completed",
        ],
        "type": "object",
      },
    },
    {
      "command": "update-todo",
      "description": "Atualiza uma tarefa existente.",
      "input_schema": {
        "properties": {
          "completed": {
            "type": "boolean",
          },
          "description": {
            "type": "string",
          },
          "id": {
            "type": "integer",
          },
          "title": {
            "type": "string",
          },
        },
        "required": [
          "id",
        ],
        "type": "object",
      },
      "name": "update-todo",
      "output_schema": {
        "properties": {
          "completed": {
            "type": "boolean",
          },
          "description": {
            "type": "string",
          },
          "id": {
            "type": "integer",
          },
          "title": {
            "type": "string",
          },
        },
        "required": [
          "id",
          "title",
          "description",
          "completed",
        ],
        "type": "object",
      },
    },
    {
      "command": "delete-todo",
      "description": "Remove uma tarefa pelo ID.",
      "input_schema": {
        "properties": {
          "id": {
            "type": "number",
          },
        },
        "required": [
          "id",
        ],
        "type": "object",
      },
      "name": "delete-todo",
      "output_schema": {
        "properties": {
          "message": {
            "type": "string",
          },
          "success": {
            "type": "boolean",
          },
        },
        "required": [
          "success",
        ],
        "type": "object",
      },
    },
    {
      "command": "get-all-todos",
      "description": "Lista todas as tarefas registradas.",
      "input_schema": {
        "properties": {},
        "type": "object",
      },
      "name": "get-all-todos",
      "output_schema": {
        "items": {
          "properties": {
            "completed": {
              "type": "boolean",
            },
            "description": {
              "type": "string",
            },
            "id": {
              "type": "number",
            },
            "title": {
              "type": "string",
            },
          },
          "required": [
            "id",
            "title",
            "description",
            "completed",
          ],
          "type": "object",
        },
        "type": "array",
      },
    },
  ],
  "description": "Gerencia tarefas com operações de CRUD completo.",
  "examples": {
    "create-todo": {
      "request": "{
  "command": "create-todo",
  "payload": {
    "title": "Estudar TypeScript",
    "description": "Praticar tipos e interfaces avançadas",
    "completed": false
  }
}",
      "response": "{
  "id": 101,
  "title": "Estudar TypeScript",
  "description": "Praticar tipos e interfaces avançadas",
  "completed": false
}",
    },
    "delete-todo": {
      "request": "{
  "command": "delete-todo",
  "payload": {
    "id": 101
  }
}",
      "response": "{
  "success": true,
  "message": "Tarefa removida com sucesso."
}",
    },
    "get-all-todos": {
      "request": "{
  "command": "get-all-todos",
  "payload": {}
}",
      "response": "[
  {
    "id": 101,
    "title": "Estudar TypeScript",
    "description": "Praticar tipos e interfaces avançadas",
    "completed": false
  },
  {
    "id": 102,
    "title": "Ler sobre Clean Architecture",
    "description": "Estudo do livro do Uncle Bob",
    "completed": true
  }
]",
    },
    "get-todo-by-id": {
      "request": "{
  "command": "get-todo-by-id",
  "payload": {
    "id": 101
  }
}",
      "response": "{
  "id": 101,
  "title": "Estudar TypeScript",
  "description": "Praticar tipos e interfaces avançadas",
  "completed": false
}",
    },
    "update-todo": {
      "request": "{
  "command": "update-todo",
  "payload": {
    "id": 101,
    "title": "Estudar TypeScript - Atualizado",
    "description": "Foco em generics e utility types",
    "completed": true
  }
}",
      "response": "{
  "id": 101,
  "title": "Estudar TypeScript - Atualizado",
  "description": "Foco em generics e utility types",
  "completed": true
}",
    },
  },
  "response": "Capacidades da ferramenta 'TodoManager':",
}
`;
